### ISSUE TITLE
RangeError: Index out of range

### ISSUE THREAD

**Issue by headshot289** [openstf/adbkit-apkreader#9]
**Error Stack:**
7/26/2017, 4:01:22 PM ERR/storage:plugins:apk 24363 [*] Unable to read manifest of "ffa8c30f-7dcb-4482-b2d1-c02fda67713c" RangeError: Index out of range
    at checkOffset (buffer.js:825:11)
    at Buffer.readUInt16LE (buffer.js:871:5)
    at BinaryXmlParser.readU16 (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/lib/apkreader/parser/binaryxml.js:98:25)
    at BinaryXmlParser.readString (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/lib/apkreader/parser/binaryxml.js:296:16)
    at BinaryXmlParser.readStringPool (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/lib/apkreader/parser/binaryxml.js:335:32)
    at BinaryXmlParser.parse (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/lib/apkreader/parser/binaryxml.js:470:12)
    at ManifestParser.parse (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/lib/apkreader/parser/manifest.js:140:33)
    at /Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/lib/apkreader.js:103:44
    at tryCatcher (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/Users/yangrenjie/gitlab/stellar-stf/node_modules/adbkit-apkreader/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:649:20)
    at tryOnImmediate (timers.js:622:5)
    at processImmediate [as _immediateCallback] (timers.js:594:5)

**AndroidManifest.xml**

```
<?xml version="1.0" encoding="utf-8"?>
<manifest
    xmlns:android="http://schemas.android.com/apk/res/android"
    android:versionCode="16"
    android:versionName="2.3.17"
    package="xxx.xxx.xxx">

    <uses-sdk
        android:minSdkVersion="14"
        android:targetSdkVersion="22" />

    <uses-permission
        android:name="android.permission.INTERNET" />
    <uses-permission
        android:name="android.permission.SYSTEM_ALERT_WINDOW" />
    <uses-permission
        android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
    <uses-permission
        android:name="android.permission.READ_PHONE_STATE" />
    <uses-permission
        android:name="android.permission.READ_EXTERNAL_STORAGE" />
    <uses-permission
        android:name="android.permission.ACCESS_WIFI_STATE" />
    <uses-permission
        android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission
        android:name="android.permission.ACCESS_COARSE_LOCATION" />
    <uses-permission
        android:name="android.permission.ACCESS_FINE_LOCATION" />
    <uses-permission
        android:name="android.permission.WRITE_SETTINGS" />

    <application
        android:theme="@ref/0x7f0d0004"
        android:label="@ref/0x7f0c0014"
        android:icon="@ref/0x7f0a0000"
        android:debuggable="true"
        android:allowBackup="true"
        android:supportsRtl="true">

        <activity
            android:name=".MainActivity">
            <intent-filter>
                <action
                    android:name="android.intent.action.MAIN" />
                <category
                    android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
    </application>
</manifest>
```

**Comment by sorccu** [comment]
I need the binary XML file from the compiled APK if you want me to look
into it. Or the whole APK.

**Comment by headshot289** [comment]
[AndroidManifest.xml.zip](https://github.com/openstf/adbkit-apkreader/files/1176283/AndroidManifest.xml.zip)

hi, this is binary file

**Comment by mingyuan-xia** [comment]
+1 for this issue, and another sample axml attached
[AndroidManifest.xml.zip](https://github.com/openstf/adbkit-apkreader/files/1230858/AndroidManifest.xml.zip)

**Comment by sorccu** [comment]
Ok, I have found the cause. Seems that we have a bug if the AndroidManifest uses utf8 encoding instead of ucs2. From what I can tell, all manifests used to be ucs2, so this issue wasn't detected. Seems that newer APKs use utf8.

**Comment by sorccu** [comment]
A fix is included in release 2.1.2.
